File: apps/delivery/src/delivery/expression/auc.py
from __future__ import annotations

import re
from pathlib import Path

import pandas as pd
import typer

from lisca.analysis.roi import load_timeseries_csv


SLIDE_CHANNEL_PATTERN = re.compile(r"_sc\d+(?=_)")
GROUP_COLUMNS = ("pos", "roi")
OUTPUT_COLUMNS = ("slide_channel", "pos", "roi", "auc")

app = typer.Typer(
    add_completion=False,
    no_args_is_help=True,
    help="Integrate ROI timeseries CSVs and write per-trace area-under-the-curve summaries.",
)


def normalize_output_stem(csv_path: Path) -> str:
    return SLIDE_CHANNEL_PATTERN.sub("", csv_path.stem)


def default_output_csv_path(timeseries_csvs: list[Path], output_csv: Path | None) -> Path:
    if output_csv is not None:
        return output_csv.resolve()

    normalized_stems = {normalize_output_stem(csv_path) for csv_path in timeseries_csvs}
    if len(normalized_stems) == 1:
        stem = next(iter(normalized_stems))
    elif len(timeseries_csvs) == 1:
        stem = timeseries_csvs[0].stem
    else:
        stem = "timeseries"
    return timeseries_csvs[0].with_name(f"{stem}_auc.csv").resolve()


def parse_slide_channel(csv_path: Path) -> int | None:
    match = re.search(r"_sc(\d+)(?=_|$)", csv_path.stem)
    if match is None:
        return None
    return int(match.group(1))


def integrate_trace(trace_df: pd.DataFrame, *, interval: float) -> float:
    sorted_df = trace_df.sort_values("t").reset_index(drop=True)
    if len(sorted_df) < 2:
        return 0.0

    times = sorted_df["t"].astype(float).to_numpy() * interval
    values = sorted_df["corrected"].astype(float).to_numpy()
    widths = times[1:] - times[:-1]
    heights = (values[:-1] + values[1:]) * 0.5
    return float((widths * heights).sum())


def compute_auc_table(timeseries_csvs: list[Path], *, interval: float) -> pd.DataFrame:
    rows: list[dict[str, object]] = []
    for csv_path in timeseries_csvs:
        df = load_timeseries_csv(csv_path)
        slide_channel = parse_slide_channel(csv_path)
        group_columns = [column for column in GROUP_COLUMNS if column in df.columns]
        if not group_columns:
            raise ValueError(f"{csv_path} has no supported grouping columns: {GROUP_COLUMNS}")

        for group_key, trace_df in df.groupby(group_columns, sort=True):
            if not isinstance(group_key, tuple):
                group_key = (group_key,)
            row = dict(zip(group_columns, group_key, strict=True))
            sorted_df = trace_df.sort_values("t").reset_index(drop=True)
            row.update(
                {
                    "slide_channel": slide_channel,
                    "auc": integrate_trace(sorted_df, interval=interval),
                }
            )
            rows.append(row)

    if not rows:
        raise ValueError("No AUC rows produced")

    result = pd.DataFrame(rows)
    sort_columns = [column for column in ("slide_channel", *GROUP_COLUMNS) if column in result.columns]
    return result.sort_values(sort_columns).reset_index(drop=True).loc[:, list(OUTPUT_COLUMNS)]


def write_auc_csv(df: pd.DataFrame, output_csv: Path) -> None:
    output_csv.parent.mkdir(parents=True, exist_ok=True)
    df.to_csv(output_csv, index=False)


@app.command()
def cli(
    timeseries_csvs: list[Path] = typer.Argument(
        ...,
        exists=True,
        dir_okay=False,
        help="One or more long-form ROI timeseries CSV files to integrate.",
    ),
    interval: float = typer.Option(
        ...,
        "--interval",
        min=0.0,
        help="Frame interval in minutes used to convert t into time before integration.",
    ),
    output_csv: Path | None = typer.Option(
        None,
        "--output-csv",
        help="Output CSV path. Default: derive a shared <stem>_auc.csv path.",
    ),
) -> None:
    if interval <= 0:
        raise ValueError(f"--interval must be > 0, got {interval}")

    resolved_csvs = sorted((csv_path.resolve() for csv_path in timeseries_csvs), key=lambda path: path.name)
    auc_df = compute_auc_table(resolved_csvs, interval=interval)
    resolved_output_csv = default_output_csv_path(resolved_csvs, output_csv)
    write_auc_csv(auc_df, resolved_output_csv)
    print(f"Wrote AUC CSV: {resolved_output_csv}")


def main(argv: list[str] | None = None, *, prog_name: str = "delivery expression auc") -> None:
    app(args=argv, prog_name=prog_name)


if __name__ == "__main__":
    main()

File: apps/delivery-rs/src/expression/auc.rs
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use clap::Args;

use lisca::analysis::roi::load_timeseries_csv;

pub const HELP: &str =
    "Integrate ROI timeseries CSVs and write per-trace area-under-the-curve summaries.";

#[derive(Clone, Debug, PartialEq)]
pub struct AucRow {
    pub slide_channel: Option<u32>,
    pub pos: Option<u32>,
    pub roi: u32,
    pub auc: f64,
}

#[derive(Clone, Debug, Args)]
#[command(about = HELP)]
pub struct AucArgs {
    #[arg(help = "One or more long-form ROI timeseries CSV files to integrate.")]
    pub timeseries_csvs: Vec<PathBuf>,
    #[arg(
        long,
        help = "Frame interval in minutes used to convert t into time before integration."
    )]
    pub interval: f64,
    #[arg(
        long,
        help = "Output CSV path. Default: derive a shared <stem>_auc.csv path."
    )]
    pub output_csv: Option<PathBuf>,
}

pub fn run_auc(
    timeseries_csvs: &[PathBuf],
    interval: f64,
    output_csv: Option<&Path>,
) -> Result<PathBuf, String> {
    if interval <= 0.0 {
        return Err(format!("--interval must be > 0, got {interval}"));
    }
    let rows = compute_auc_table(timeseries_csvs, interval)?;
    let output_path = default_output_csv_path(timeseries_csvs, output_csv);
    write_auc_csv(&rows, &output_path)?;
    Ok(output_path)
}

pub fn normalize_output_stem(csv_path: &Path) -> String {
    let stem = csv_path
        .file_stem()
        .and_then(|value| value.to_str())
        .unwrap_or_default();
    normalize_output_stem_with(stem, false)
}

pub fn parse_image_channel(csv_path: &Path) -> Option<u32> {
    let stem = csv_path.file_stem()?.to_str()?;
    let marker = stem.find("_ch")?;
    let digits = stem[marker + 3..]
        .chars()
        .take_while(|c| c.is_ascii_digit())
        .collect::<String>();
    if digits.is_empty() {
        return None;
    }
    digits.parse().ok()
}

pub fn aggregate_output_stem(timeseries_csvs: &[PathBuf]) -> String {
    let parsed_channels = timeseries_csvs
        .iter()
        .filter_map(|path| parse_image_channel(path))
        .collect::<std::collections::BTreeSet<_>>();
    let drop_image_channel = parsed_channels.len() > 1;
    let stems = timeseries_csvs
        .iter()
        .map(|path| {
            let stem = path
                .file_stem()
                .and_then(|value| value.to_str())
                .unwrap_or_default();
            normalize_output_stem_with(stem, drop_image_channel)
        })
        .collect::<std::collections::BTreeSet<_>>();
    if stems.len() == 1 {
        stems.into_iter().next().unwrap()
    } else if timeseries_csvs.len() == 1 {
        let stem = timeseries_csvs[0]
            .file_stem()
            .and_then(|value| value.to_str())
            .unwrap_or_default();
        normalize_output_stem_with(stem, drop_image_channel)
    } else {
        "timeseries".to_string()
    }
}

pub fn aggregate_output_stem_candidates(csv_path: &Path) -> std::collections::BTreeSet<String> {
    let stem = csv_path
        .file_stem()
        .and_then(|value| value.to_str())
        .unwrap_or_default();
    std::collections::BTreeSet::from([
        normalize_output_stem_with(stem, false),
        normalize_output_stem_with(stem, true),
    ])
}

pub fn default_output_csv_path(timeseries_csvs: &[PathBuf], output_csv: Option<&Path>) -> PathBuf {
    if let Some(path) = output_csv {
        return path.to_path_buf();
    }

    let stem = aggregate_output_stem(timeseries_csvs);
    timeseries_csvs
        .first()
        .cloned()
        .unwrap_or_else(|| PathBuf::from("timeseries.csv"))
        .with_file_name(format!("{stem}_auc.csv"))
}

pub fn parse_slide_channel(csv_path: &Path) -> Option<u32> {
    let stem = csv_path.file_stem()?.to_str()?;
    let marker = stem.find("_sc")?;
    let digits = stem[marker + 3..]
        .chars()
        .take_while(|c| c.is_ascii_digit())
        .collect::<String>();
    if digits.is_empty() {
        return None;
    }
    digits.parse().ok()
}

pub fn integrate_trace(trace: &[(u32, f64)], interval: f64) -> f64 {
    if trace.len() < 2 {
        return 0.0;
    }

    trace
        .windows(2)
        .map(|window| {
            let (t0, v0) = window[0];
            let (t1, v1) = window[1];
            let width = (t1 as f64 - t0 as f64) * interval;
            let height = (v0 + v1) * 0.5;
            width * height
        })
        .sum()
}

pub fn compute_auc_table(
    timeseries_csvs: &[PathBuf],
    interval: f64,
) -> Result<Vec<AucRow>, String> {
    let mut auc_rows = Vec::new();
    let mut csvs = timeseries_csvs.to_vec();
    csvs.sort_by(|a, b| a.file_name().cmp(&b.file_name()));

    for csv_path in csvs {
        let slide_channel = parse_slide_channel(&csv_path);
        let rows = load_timeseries_csv(&csv_path)?;
        let mut grouped = BTreeMap::<(Option<u32>, u32), Vec<(u32, f64)>>::new();
        for row in rows {
            grouped
                .entry((row.pos, row.roi))
                .or_default()
                .push((row.t, row.corrected));
        }
        for ((pos, roi), mut trace) in grouped {
            trace.sort_by_key(|(t, _)| *t);
            auc_rows.push(AucRow {
                slide_channel,
                pos,
                roi,
                auc: integrate_trace(&trace, interval),
            });
        }
    }

    if auc_rows.is_empty() {
        return Err("No AUC rows produced".to_string());
    }

    auc_rows.sort_by(|a, b| {
        a.slide_channel
            .cmp(&b.slide_channel)
            .then(a.pos.cmp(&b.pos))
            .then(a.roi.cmp(&b.roi))
    });
    Ok(auc_rows)
}

pub fn write_auc_csv(rows: &[AucRow], output_csv: &Path) -> Result<(), String> {
    if let Some(parent) = output_csv.parent() {
        std::fs::create_dir_all(parent).map_err(|err| err.to_string())?;
    }

    let mut writer = csv::Writer::from_path(output_csv).map_err(|err| err.to_string())?;
    writer
        .write_record(["slide_channel", "pos", "roi", "auc"])
        .map_err(|err| err.to_string())?;
    for row in rows {
        writer
            .write_record([
                row.slide_channel
                    .map(|value| value.to_string())
                    .unwrap_or_default(),
                row.pos.map(|value| value.to_string()).unwrap_or_default(),
                row.roi.to_string(),
                row.auc.to_string(),
            ])
            .map_err(|err| err.to_string())?;
    }
    writer.flush().map_err(|err| err.to_string())
}

pub fn format_written_auc_csv_message(output_csv: &Path) -> String {
    format!("Wrote AUC CSV: {}", output_csv.display())
}

pub fn execute(args: AucArgs) -> Result<(), String> {
    let output = run_auc(
        &args.timeseries_csvs,
        args.interval,
        args.output_csv.as_deref(),
    )?;
    println!("{}", format_written_auc_csv_message(&output));
    Ok(())
}

fn remove_slide_channel_segment(stem: &str) -> String {
    if let Some(marker) = stem.find("_sc") {
        let digits_end = stem[marker + 3..]
            .chars()
            .take_while(|c| c.is_ascii_digit())
            .count();
        if digits_end > 0 {
            let end = marker + 3 + digits_end;
            if stem.as_bytes().get(end) == Some(&b'_') {
                let mut normalized = stem[..marker].to_string();
                normalized.push_str(&stem[end..]);
                return normalized;
            }
        }
    }
    stem.to_string()
}

fn remove_image_channel_segment(stem: &str) -> String {
    if let Some(marker) = stem.find("_ch") {
        let digits_end = stem[marker + 3..]
            .chars()
            .take_while(|c| c.is_ascii_digit())
            .count();
        if digits_end > 0 {
            let end = marker + 3 + digits_end;
            if stem.as_bytes().get(end) == Some(&b'_') {
                let mut normalized = stem[..marker].to_string();
                normalized.push_str(&stem[end..]);
                return normalized;
            }
        }
    }
    stem.to_string()
}

fn normalize_output_stem_with(stem: &str, drop_image_channel: bool) -> String {
    let normalized = remove_slide_channel_segment(stem);
    if drop_image_channel {
        remove_image_channel_segment(&normalized)
    } else {
        normalized
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_output_csv_path_strips_slide_channel_segment() {
        let csv_paths = vec![
            PathBuf::from("/tmp/slide_sc0_ch001_timeseries.csv"),
            PathBuf::from("/tmp/slide_sc2_ch001_timeseries.csv"),
        ];
        let output = default_output_csv_path(&csv_paths, None);
        assert!(output.ends_with("slide_ch001_timeseries_auc.csv"));
    }

    #[test]
    fn default_output_csv_path_drops_image_channel_for_mixed_inputs() {
        let csv_paths = vec![
            PathBuf::from("/tmp/slide_sc0_ch001_timeseries.csv"),
            PathBuf::from("/tmp/slide_sc2_ch002_timeseries.csv"),
        ];
        let output = default_output_csv_path(&csv_paths, None);
        assert!(output.ends_with("slide_timeseries_auc.csv"));
    }

    #[test]
    fn integrate_trace_uses_trapezoidal_rule() {
        let trace = vec![(0, 1.0), (1, 3.0), (2, 5.0)];
        assert_eq!(integrate_trace(&trace, 10.0), 60.0);
    }
}
